### PR TITLE
Add a trigger migration to enable a bug in ABCI error decoding

### PR DESCRIPTION
**Pull request by hansl**
This bug is needed to allow legacy behaviour while replaying old blocks. It requires a new migration type to allow for modern behaviour to be enabled by default, which we want moving forward.

### CODE REVIEW

**Review by fmorency — COMMENTED** [review]
Thanks!
I "fixed" your resiliency test, but you'll need to make additional changes for it to work; see below.

File: tests/resiliency/ledger/BUILD.bazel [hunk 1]
```diff
@@ -4,6 +4,7 @@ bats_test(
     name = "bats-resiliency-ledger",
     srcs = [
         "abci-allow-addrs.bats",
+        "abci_migration.bats",
```
Comment (fmorency) [inline tests/resiliency/ledger/BUILD.bazel:7]: FYI, this is improved in https://github.com/liftedinit/many-rs/pull/320

File: scripts/run.sh [hunk 2]
```diff
@@ -165,12 +165,10 @@ main() {
...
-    # This makes sure the sessions remain opened when the command exits.
-    tmux setw remain-on-exit on
+    tmux new-window -t "$tmux_name" -n tendermint-kvstore -e SHELL=bash "TMHOME=\"$root_dir/kvstore\" tendermint start 2>&1 | tee \"$root_dir/tendermint-kvstore.log\""
 
     tmux new-window -t "$tmux_name" -n ledger -e SHELL=bash "./target/debug/many-ledger -v -v --abci --addr 127.0.0.1:8001 --pem \"$pem_root/id1.pem\" --state ./staging/ledger_state.json5 --persistent \"$root_dir/ledger.db\" 2>&1 | tee \"$root_dir/many-ledger.log\""
-    tmux new-window -t "$tmux_name" -n ledger-abci "./target/debug/many-abci -v -v --many 127.0.0.1:8000 --many-app http://localhost:8001 --many-pem \"$pem_root/id2.pem\" --abci 127.0.0.1:26658 --tendermint http://127.0.0.1:26657/ 2>&1 | tee \"$root_dir/many-abci-ledger.log\""
+    tmux new-window -t "$tmux_name" -n ledger-abci "./target/debug/many-abci -v -v --many 127.0.0.1:8000 --many-app http://localhost:8001 --many-pem \"$pem_root/id2.pem\" --abci 127.0.0.1:26658 --tendermint http://127.0.0.1:26657/ --migrations-config=\"./tests/abci_migrations.json\" 2>&1 | tee \"$root_dir/many-abci-ledger.log\""
```
Comment (fmorency) [inline scripts/run.sh:171]: I think you can revert those changes.

File: src/many-migration/src/lib.rs [hunk 3]
```diff
@@ -182,11 +210,17 @@ impl<T, E> InnerMigration<T, E> {
...
+    pub const fn is_trigger(&self) -> bool {
+        matches!(self.r#type, MigrationType::Trigger(_))
+    }
+
     /// This function gets executed when the storage block height == the migration block height
     fn initialize(&self, storage: &mut T, extra: &HashMap<String, Value>) -> Result<(), E> {
         match &self.r#type {
             MigrationType::Regular(migration) => (migration.initialize_fn)(storage, extra),
             MigrationType::Hotfix(_) => Ok(()),
+            MigrationType::Trigger(_) => Ok(()),
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:223]: nit: use `|`?

File: src/many-migration/src/lib.rs [hunk 4]
```diff
@@ -199,6 +233,7 @@ impl<T, E> InnerMigration<T, E> {
         match &self.r#type {
             MigrationType::Regular(migration) => (migration.update_fn)(storage, extra),
             MigrationType::Hotfix(_) => Ok(()),
+            MigrationType::Trigger(_) => Ok(()),
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:236]: nit: use `|`?

File: src/many-migration/src/lib.rs [hunk 5]
```diff
@@ -211,6 +246,7 @@ impl<T, E> InnerMigration<T, E> {
         match &self.r#type {
             MigrationType::Regular(_) => None,
             MigrationType::Hotfix(migration) => (migration.hotfix_fn)(b),
+            MigrationType::Trigger(_) => None,
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:249]: nit: use `|`?

File: src/many-migration/src/lib.rs [hunk 6]
```diff
@@ -267,17 +303,37 @@ impl<'a, T, E> Migration<'a, T, E> {
         }
     }
 
+    pub fn activate_at_height(&mut self, height: u64) -> (bool, bool) {
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:306]: Not clear what the return value of this function is. Add a comment.

File: src/many-migration/tests/migrations.rs [hunk 7]
```diff
@@ -1,465 +0,0 @@
-#![feature(used_with_arg)] // Required to build the test with Bazel
```
Comment (fmorency) [inline src/many-migration/tests/migrations.rs:1]: Why did you remove those tests? Where are those tested now?

File: tests/resiliency/ledger/abci_migration.bats [hunk 8]
```diff
@@ -0,0 +1,67 @@
+GIT_ROOT="$BATS_TEST_DIRNAME/../../../"
+#MIGRATION_ROOT="$GIT_ROOT/tests/ledger_migrations.json"
```
Comment (fmorency) [inline tests/resiliency/ledger/abci_migration.bats:2]: Remove

**Review by fmorency — CHANGES_REQUESTED** [review]
Thanks!

File: src/many-migration/tests/migrations.rs (hunk 7 above)
Comment (hansl) [reply src/many-migration/tests/migrations.rs:1]: My mistake, I deleted the wrong files.

**Review by fmorency — APPROVED** [review]

**Comment by fmorency** [comment]
Assuming CI passes, LGTM!
